documentation/src/docs/controller-service.md

```diff
@@ -544,6 +529,57 @@ Streams to go on concurrently. Workflows like _scale, truncation, seal,
+
+## Controller Cluster Listener
+
+Each node in Pravega Cluster registers itself under a cluster znode as
+an ephemeral node. This includes both Controller and Segment Store
+nodes. All segment stores register under a child znode called hosts while all controllers register under a child znode called controllers. 
+Controller Cluster Listener is the component that monitors the cluster of controllers and is chiefly responsible for ensuring that any partially
+completed work from a failed controller instance is completed by surviving controller instances. To achieve this each Controller instance registers a watch on the controller cluster znode
+to listen for cluster change notifications. These notify about the added and removed nodes.
+
+All Controller instances listen for Controller node change
+notifications. Each Controller instance has multiple sub components that
+implement the _failover sweeper_ interface. The purpose of failover sweeper is to sweep any unfinished ongoing tasks from the failed controller instance. 
+Any controller sub-component that wishes to implement a failover logic should implement the FailoverSweeper interface. 
+Presently there are three components that implement _failover_ sweeper interface namely:
+
+- `TaskSweeper`
+- `EventProcessors`
+- `TransactionSweeper`
+
+Whenever a Controller instance is identified to have been removed from the cluster,
+the cluster listener invokes all registered _failover sweepers_ to
+optimistically try to sweep all the orphaned work previously owned by
+the failed Controller host.
+
+TaskSweeper is for the task framework which allows running background workflows on a resource (typically a stream) by acqiring a zookeeper based distributed lock for mutual exclusion. 
+All ongoing tasks that a controller wishes to acquire, it first adds an entry into its index about it before attempting to acquire the distributed lock. Upon successful acquisition, the workflow is performed, else the indexed entry is removed. The indexed entry is only an indicator that this controller may attempt to do the said work. The truth of acquisition is on the distributed lock. If the controller fails with an indexed entry then any other controller attempting to sweep its tasks would look at all the indexed entries and check the ownership for those tasks by the failed controller. If the failed controller did indeed own those tasks, then the new controller will attempt to first create an index and then acquire the lock. 
+
+Just like task sweeper and request sweepers also rely on creating an index in zookeeper to indicate an intent to atomically update a metadata record and post an event into the respective event processor stream. If the controller fails after updating the metadata but without posting an event into the event processor stream, then the new controller will sweep all such indexed tasks from zookeeper and post the events for them. 
```

Comment:
Here `controller` is written again without capital C. Also, Zookeeper with capital Z.